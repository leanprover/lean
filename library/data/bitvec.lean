/-
Copyright (c) 2015 Joe Hendrix. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Joe Hendrix, Sebastian Ullrich

Basic operations on bitvectors.

This is a work-in-progress, and contains additions to other theories.
-/
import data.tuple

@[reducible] def bitvec (n : ℕ) := tuple bool n

namespace bitvec
open nat
open tuple

local infix `++ₜ`:65 := tuple.append

-- Create a zero bitvector
@[reducible] protected def zero (n : ℕ) : bitvec n := repeat ff n

-- Create a bitvector with the constant one.
@[reducible] protected def one : Π (n : ℕ), bitvec n
| 0        := []
| (succ n) := repeat ff n ++ₜ [tt]

protected def cong {a b : ℕ} (h : a = b) : bitvec a → bitvec b
| ⟨x, p⟩ := ⟨x, h ▸ p⟩

section shift
  variable {n : ℕ}

  def shl (x : bitvec n) (i : ℕ) : bitvec n :=
  let r := dropn i x ++ₜ repeat ff (min n i) in
  have length r = n, begin dsimp, rewrite [nat.sub_add_min_cancel] end,
  bitvec.cong this r

  def fill_shr (x : bitvec n) (i : ℕ) (fill : bool) : bitvec n :=
  let y := repeat fill (min n i) ++ₜ firstn (n-i) x in
  have length y = n, from if h : i ≤ n then
    begin
      dsimp,
      rw [min_eq_right h, min_eq_left (sub_le _ _), -nat.add_sub_assoc h,
        nat.add_sub_cancel_left]
    end
  else
    have h : i ≥ n, from le_of_not_ge h,
    begin
      dsimp,
      rw [min_eq_left h, sub_eq_zero_of_le h, min_eq_left (zero_le _)],
      apply rfl
    end,
  bitvec.cong this y

  -- unsigned shift right
  def ushr (x : bitvec n) (i : ℕ) : bitvec n :=
  fill_shr x i ff

  -- signed shift right
  def sshr : Π {m : ℕ}, bitvec m → ℕ → bitvec m
  | 0        _ _ := []
  | (succ m) x i := head x :: fill_shr (tail x) i (head x)

end shift

section bitwise
  variable {n : ℕ}

  def not : bitvec n → bitvec n := map bnot
  def and : bitvec n → bitvec n → bitvec n := map₂ band
  def or  : bitvec n → bitvec n → bitvec n := map₂ bor
  def xor : bitvec n → bitvec n → bitvec n := map₂ bxor

end bitwise

section arith
  variable {n : ℕ}

  protected def xor3 (x y c : bool) := bxor (bxor x y) c
  protected def carry (x y c : bool) :=
  x && y || x && c || y && c

  def neg (x : bitvec n) : bitvec n :=
  let f := λ y c, (y || c, bxor y c) in
  prod.snd (map_accumr f x ff)

  -- Add with carry (no overflow)
  def adc (x y : bitvec n) (c : bool) : bitvec (n+1) :=
  let f := λ x y c, (bitvec.carry x y c, bitvec.xor3 x y c) in
  let ⟨c, z⟩ := tuple.map_accumr₂ f x y c in
  c :: z

  def add (x y : bitvec n) : bitvec n := tail (adc x y ff)

  protected def borrow (x y b : bool) :=
  bnot x && y || bnot x && b || y && b

  -- Subtract with borrow
  def sbb (x y : bitvec n) (b : bool) : bool × bitvec n :=
  let f := λ x y c, (bitvec.borrow x y c, bitvec.xor3 x y c) in
  tuple.map_accumr₂ f x y b

  def sub (x y : bitvec n) : bitvec n := prod.snd (sbb x y ff)

  instance : has_zero (bitvec n) := ⟨bitvec.zero n⟩
  instance : has_one (bitvec n)  := ⟨bitvec.one n⟩
  instance : has_add (bitvec n)  := ⟨add⟩
  instance : has_sub (bitvec n)  := ⟨sub⟩
  instance : has_neg (bitvec n)  := ⟨neg⟩

  def mul (x y : bitvec n) : bitvec n :=
  let f := λ r b, cond b (r + r + y) (r + r) in
  list.foldl f 0 (to_list x)

  instance : has_mul (bitvec n)  := ⟨mul⟩
end arith

section comparison
  variable {n : ℕ}

  def uborrow (x y : bitvec n) : bool := prod.fst (sbb x y ff)

  def ult (x y : bitvec n) : Prop := uborrow x y
  def ugt (x y : bitvec n) : Prop := ult y x

  def ule (x y : bitvec n) : Prop := ¬ (ult y x)
  def uge (x y : bitvec n) : Prop := ule y x

  def sborrow : Π {n : ℕ}, bitvec n → bitvec n → bool
  | 0        _ _ := ff
  | (succ n) x y :=
     bool.cases_on
       (head x)
       (bool.cases_on (head y) (uborrow (tail x) (tail y)) tt)
       (bool.cases_on (head y) ff (uborrow (tail x) (tail y)))

  def slt (x y : bitvec n) : Prop := sborrow x y
  def sgt (x y : bitvec n) : Prop := slt y x
  def sle (x y : bitvec n) : Prop := ¬ (slt y x)
  def sge (x y : bitvec n) : Prop := sle y x

end comparison

section conversion
  variable {α : Type}

  protected def of_nat : Π (n : ℕ), nat → bitvec n
  | 0        x := nil
  | (succ n) x := of_nat n (x / 2) ++ₜ [to_bool (x % 2 = 1)]

  protected def of_int : Π (n : ℕ), int → bitvec (succ n)
  | n (int.of_nat m)          := ff :: bitvec.of_nat n m
  | n (int.neg_succ_of_nat m) := tt :: not (bitvec.of_nat n m)

  def bits_to_nat (v : list bool) : nat :=
  list.foldl (λ r b, r + r + cond b 1 0) 0 v

  protected def to_nat {n : nat} (v : bitvec n) : nat :=
  bits_to_nat (to_list v)

  protected def to_int : Π {n : nat}, bitvec n → int
  | 0        _ := 0
  | (succ n) v :=
    cond (head v)
      (int.neg_succ_of_nat $ bitvec.to_nat $ not $ tail v)
      (int.of_nat $ bitvec.to_nat $ tail v)

end conversion

private def to_string {n : nat} : bitvec n → string
| ⟨bs, p⟩ :=
  "0b" ++ (bs^.reverse^.for (λ b, if b then #"1" else #"0"))

instance (n : nat) : has_to_string (bitvec n) :=
⟨to_string⟩
end bitvec
