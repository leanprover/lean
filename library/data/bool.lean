/-
Copyright (c) 2014 Microsoft Corporation. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Author: Leonardo de Moura
-/

import logic.eq
open eq eq.ops decidable

namespace bool
  local attribute bor [reducible]
  local attribute band [reducible]

  theorem dichotomy (b : bool) : b = ff ∨ b = tt :=
  bool.cases_on b (or.inl rfl) (or.inr rfl)

  theorem cond_ff {A : Type} (t e : A) : cond ff t e = e :=
  rfl

  theorem cond_tt {A : Type} (t e : A) : cond tt t e = t :=
  rfl

  theorem eq_tt_of_ne_ff : ∀ {a : bool}, a ≠ ff → a = tt
  | @@eq_tt_of_ne_ff tt H := rfl
  | @@eq_tt_of_ne_ff ff H := absurd rfl H

  theorem eq_ff_of_ne_tt : ∀ {a : bool}, a ≠ tt → a = ff
  | @@eq_ff_of_ne_tt tt H := absurd rfl H
  | @@eq_ff_of_ne_tt ff H := rfl

  theorem absurd_of_eq_ff_of_eq_tt {B : Prop} {a : bool} (H₁ : a = ff) (H₂ : a = tt) : B :=
  absurd (H₁⁻¹ ⬝ H₂) ff_ne_tt

  theorem tt_bor (a : bool) : bor tt a = tt :=
  rfl

  notation a || b := bor a b

  theorem bor_tt (a : bool) : a || tt = tt :=
  bool.cases_on a rfl rfl

  theorem ff_bor (a : bool) : ff || a = a :=
  bool.cases_on a rfl rfl

  theorem bor_ff (a : bool) : a || ff = a :=
  bool.cases_on a rfl rfl

  theorem bor_self (a : bool) : a || a = a :=
  bool.cases_on a rfl rfl

  theorem bor.comm (a b : bool) : a || b = b || a :=
  by cases a; repeat (cases b | reflexivity)

  theorem bor.assoc (a b c : bool) : (a || b) || c = a || (b || c) :=
  match a with
  | ff := by rewrite *ff_bor
  | tt := by rewrite *tt_bor
  end

  theorem or_of_bor_eq {a b : bool} : a || b = tt → a = tt ∨ b = tt :=
  bool.rec_on a
    (suppose ff || b = tt,
      have b = tt, from !ff_bor ▸ this,
      or.inr this)
    (suppose tt || b = tt,
      or.inl rfl)

  theorem bor_inl {a b : bool} (H : a = tt) : a || b = tt :=
  by rewrite H

  theorem bor_inr {a b : bool} (H : b = tt) : a || b = tt :=
  bool.rec_on a (by rewrite H) (by rewrite H)

  theorem ff_band (a : bool) : ff && a = ff :=
  rfl

  theorem tt_band (a : bool) : tt && a = a :=
  bool.cases_on a rfl rfl

  theorem band_ff (a : bool) : a && ff = ff :=
  bool.cases_on a rfl rfl

  theorem band_tt (a : bool) : a && tt = a :=
  bool.cases_on a rfl rfl

  theorem band_self (a : bool) : a && a = a :=
  bool.cases_on a rfl rfl

  theorem band.comm (a b : bool) : a && b = b && a :=
  bool.cases_on a
    (bool.cases_on b rfl rfl)
    (bool.cases_on b rfl rfl)

  theorem band.assoc (a b c : bool) : (a && b) && c = a && (b && c) :=
  match a with
  | ff := by rewrite *ff_band
  | tt := by rewrite *tt_band
  end

  theorem band_elim_left {a b : bool} (H : a && b = tt) : a = tt :=
  or.elim (dichotomy a)
    (suppose a = ff,
      absurd
        (calc ff = ff && b : ff_band
             ... = a && b  : this
             ... = tt      : H)
        ff_ne_tt)
    (suppose a = tt, this)

  theorem band_intro {a b : bool} (H₁ : a = tt) (H₂ : b = tt) : a && b = tt :=
  by rewrite [H₁, H₂]

  theorem band_elim_right {a b : bool} (H : a && b = tt) : b = tt :=
  band_elim_left (!band.comm ⬝ H)

  theorem bnot_bnot (a : bool) : bnot (bnot a) = a :=
  bool.cases_on a rfl rfl

  theorem bnot_false : bnot ff = tt :=
  rfl

  theorem bnot_true  : bnot tt = ff :=
  rfl

  theorem eq_tt_of_bnot_eq_ff {a : bool} : bnot a = ff → a = tt :=
  bool.cases_on a (by contradiction) (λ h, rfl)

  theorem eq_ff_of_bnot_eq_tt {a : bool} : bnot a = tt → a = ff :=
  bool.cases_on a (λ h, rfl) (by contradiction)
end bool
