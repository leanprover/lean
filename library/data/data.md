data
====

Various data types.

Basic types:

* [empty](empty.lean) : the empty type
* [unit](unit.lean) : the singleton type
* [bool](bool.lean) : the boolean values
* [num](num.lean) : generic numerals
* [string](string.lean) : ascii strings
* [nat](nat/nat.md) : the natural numbers
* [fin](fin.lean) : finite ordinals
* [less_than](less_than.lean) : finite ordinals
* [int](int/int.md) : the integers
* [rat](rat/rat.md) : the rationals

Constructors:

* [prod](prod.lean) : cartesian product
* [sum](sum.lean)
* [sigma](sigma.lean) : the dependent product
* [uprod](uprod.lean) : unordered pairs
* [option](option.lean)
* [subtype](subtype.lean)
* [quotient](quotient/quotient.md)
* [squash](squash.lean) : propositional truncation
* [list](list/list.md)
* [finset](finset/finset.md) : finite sets
* [stream](stream.lean)
* [set](set/set.md)
* [vector](vector.lean)

Types with extra information:

* [fintype](fintype/fintype.md) : finite types
* [encodable](encodable.lean) : types with a coding to nat
