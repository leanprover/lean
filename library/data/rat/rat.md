data.rat
========

The rational numbers.

* [basic](basic.lean) : the rationals as a field
* [order](order.lean) : the order relations and the sign function
