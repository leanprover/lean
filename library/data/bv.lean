/-
Copyright (c) 2015 Joe Hendrix. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Author: Joe Hendrix

Basic operations on bitvectors.

This is a work-in-progress, and contains additions to other theories.
-/
import data.list
import data.tuple

namespace bv
open bool
open eq.ops
open list
open nat
open prod
open subtype
open tuple

attribute [reducible]
definition bv (n : ℕ) := tuple bool n

-- Create a zero bitvector
definition bv_zero (n : ℕ) : bv n := replicate ff

-- Create  a bitvector with the constant one.
definition bv_one  : Π (n : ℕ), bv n
  | 0 := replicate ff
  | (succ n) := (replicate ff : bv n) ++ (tt :: nil)

definition bv_cong {a b : ℕ} : (a = b) → bv a → bv b
| c (tag x p) := tag x (c ▸ p)

section shift

  -- shift left
  definition bv_shl {n:ℕ} : bv n → ℕ → bv n
    | x i :=
       if le : i ≤ n then
         let r := dropn i x ++ replicate ff in
         let eq := calc (n-i) + i = n : nat.sub_add_cancel le in
         bv_cong eq r
       else
         bv_zero n

  -- unsigned shift right
  definition bv_ushr {n:ℕ} : bv n → ℕ → bv n
    | x i :=
      if le : i ≤ n then
        let y : bv (n-i) := @firstn _ _ (n - i) (sub_le n i) x in
        let eq := calc (i+(n-i)) = (n - i) + i : add.comm
                            ...  = n           : nat.sub_add_cancel le in
        bv_cong eq (replicate ff ++ y)
      else
        bv_zero n

  -- signed shift right
  definition bv_sshr {m:ℕ} : bv (succ m) → ℕ → bv (succ m)
    | x i :=
      let n := succ m in
      if le : i ≤ n then
        let z : bv i := replicate (head x) in
        let y : bv (n-i) := @firstn _ _ (n - i) (sub_le n i) x in
        let eq := calc (i+(n-i)) = (n-i) + i : add.comm
                            ...  = n         : nat.sub_add_cancel le in
        bv_cong eq (z ++ y)
      else
        bv_zero n

end shift

section bitwise
  variable { n : ℕ }

  definition bv_not : bv n → bv n := map bnot
  definition bv_and : bv n → bv n → bv n := map₂ band
  definition bv_or  : bv n → bv n → bv n := map₂ bor
  definition bv_xor : bv n → bv n → bv n := map₂ bxor

end bitwise

section arith

  variable { n : ℕ }

  protected definition xor3 (x:bool) (y:bool) (c:bool) := bxor (bxor x y) c
  protected definition carry (x:bool) (y:bool) (c:bool) :=
    x && y || x && c || y && c

  definition bv_neg : bv n → bv n
  | x :=
    let f := λy c, (y || c, bxor y c) in
    pr₂ (mapAccumR f x ff)

  -- Add with carry (no overflow)
  definition bv_adc : bv n → bv n → bool → bv (n+1)
  | x y c :=
    let f := λx y c, (bv.carry x y c, bv.xor3 x y c) in
    let z := tuple.mapAccumR₂ f x y c in
    (pr₁ z) :: (pr₂ z)

  definition bv_add : bv n → bv n → bv n
  | x y := tail (bv_adc x y ff)

  protected definition borrow (x:bool) (y:bool) (b:bool) :=
    bnot x && y || bnot x && b || y && b

  -- Subtract with borrow
  definition bv_sbb : bv n → bv n → bool → bool × bv n
  | x y b :=
    let f := λx y c, (bv.borrow x y c, bv.xor3 x y c) in
    tuple.mapAccumR₂ f x y b

  definition bv_sub : bv n → bv n → bv n
  | x y := pr₂ (bv_sbb x y ff)

  attribute [instance]
  definition bv_has_zero : has_zero (bv n) := has_zero.mk (bv_zero n)
  attribute [instance]
  definition bv_has_one  : has_one (bv n)  := has_one.mk (bv_one n)
  attribute [instance]
  definition bv_has_add  : has_add (bv n)  := has_add.mk bv_add
  attribute [instance]
  definition bv_has_sub  : has_sub (bv n)  := has_sub.mk bv_sub
  attribute [instance]
  definition bv_has_neg  : has_neg (bv n)  := has_neg.mk bv_neg

  definition bv_mul : bv n → bv n → bv n
  | x y :=
    let f := λr b, cond b (r + r + y) (r + r) in
    foldl f 0 (to_list x)

  attribute [instance]
  definition bv_has_mul  : has_mul (bv n)  := has_mul.mk bv_mul

  definition bv_ult : bv n → bv n → bool := λx y, pr₁ (bv_sbb x y ff)
  definition bv_ugt : bv n → bv n → bool := λx y, bv_ult y x
  definition bv_ule : bv n → bv n → bool := λx y, bnot (bv_ult y x)
  definition bv_uge : bv n → bv n → bool := λx y, bv_ule y x

  definition bv_slt : bv (succ n) → bv (succ n) → bool := λx y,
    cond (head x)
         (cond (head y)
               (bv_ult (tail x) (tail y))  -- both negative
               tt)                         -- x is negative and y is not
         (cond (head y)
               ff                          -- y is negative and x is not
               (bv_ult (tail x) (tail y))) -- both positive
  definition bv_sgt : bv (succ n) → bv (succ n) → bool := λx y, bv_slt y x
  definition bv_sle : bv (succ n) → bv (succ n) → bool := λx y, bnot (bv_slt y x)
  definition bv_sge : bv (succ n) → bv (succ n) → bool := λx y, bv_sle y x
end arith


section from_bv
  variable {A : Type}

  -- Convert a bitvector to another number.
  definition from_bv [p : has_add A] [q0 : has_zero A] [q1 : has_one A] {n:nat} (v:bv n) : A :=
    let f := λr b, cond b (r + r + 1) (r + r) in
    foldl f 0 (to_list v)
end from_bv

end bv
