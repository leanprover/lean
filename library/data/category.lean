-- category

import logic.core.eq logic.core.connectives
import data.unit data.sigma data.prod
import struc.function

inductive category (ob : Type) (mor : ob → ob → Type) : Type :=
cat_mk : Π (comp : Π⦃A B C : ob⦄, mor B C → mor A B → mor A C)
           (id : Π {A : ob}, mor A A),
            (Π {A B C D : ob} {f : mor A B} {g : mor B C} {h : mor C D},
            comp h (comp g f) = comp (comp h g) f) →
           (Π {A B : ob} {f : mor A B}, comp f id = f) →
           (Π {A B : ob} {f : mor A B}, comp id f = f) →
            category ob mor

namespace category

section

parameters {ob : Type} {mor : ob → ob → Type} {Cat : category ob mor}


abbreviation compose := category_rec (λ comp id assoc idr idl, comp) Cat

precedence `∘`:60
infixr  ∘ := compose

abbreviation id := category_rec (λ comp id assoc idr idl, id) Cat
abbreviation ID (A : ob) := @id A

theorem assoc : Π {A B C D : ob} {f : mor A B} {g : mor B C} {h : mor C D},
            h ∘ (g ∘ f) = (h ∘ g) ∘ f :=
category_rec (λ comp id assoc idr idl, assoc) Cat
theorem id_right : Π {A B : ob} {f : mor A B}, f ∘ id = f :=
category_rec (λ comp id assoc idr idl, idr) Cat
theorem id_left  : Π {A B : ob} {f : mor A B}, id ∘ f = f :=
category_rec (λ comp id assoc idr idl, idl) Cat

theorem left_id_unique {A : ob} (i : mor A A) (H : Π{B} {f : mor B A}, i ∘ f = f) : i = id :=
calc
  i = i ∘ id : symm id_right
  ... = id : H

theorem right_id_unique {A : ob} (i : mor A A) (H : Π{B} {f : mor A B}, f ∘ i = f) : i = id :=
calc
  i = id ∘ i : symm id_left
  ... = id : H

definition iso {A B : ob} (f : mor A B) : Type := Σ g, f ∘ g = id ∧ g ∘ f = id
definition inverse {A B : ob} (f : mor A B) (H : iso f) : mor B A := sigma.dpr1 H

postfix `⁻¹` := inverse

theorem compose_inverse {A B : ob} {f : mor A B} (H : iso f) : f ∘ f⁻¹ H = id :=
and_elim_left (sigma.dpr2 H)

theorem inverse_compose {A B : ob} {f : mor A B} (H : iso f) : f⁻¹ H ∘ f = id :=
and_elim_right (sigma.dpr2 H)

theorem inverse_unique {A B : ob} {f : mor A B} (H H' : iso f) : f⁻¹ H = f⁻¹ H' :=
calc
  f⁻¹ H = f⁻¹ H ∘ id : symm id_right
    ... = f⁻¹ H ∘ f ∘ f⁻¹ H' : {symm (compose_inverse H')}
    ... = (f⁻¹ H ∘ f) ∘ f⁻¹ H' : assoc
    ... = id ∘ f⁻¹ H' : {inverse_compose H}
    ... = f⁻¹ H' : id_left

definition mono {A B : ob} (f : mor A B) : Prop := ∀⦃C⦄ {g h : mor C A}, f ∘ g = f ∘ h → g = h
definition epi  {A B : ob} (f : mor A B) : Prop := ∀⦃C⦄ {g h : mor B C}, g ∘ f = h ∘ f → g = h

end


using unit

definition one [instance] : category unit (λa b, unit) :=
cat_mk (λ a b c f g, star) (λ a, star) (λ a b c d f g h, unit_eq _ _)
  (λ a b f, unit_eq _ _) (λ a b f, unit_eq _ _)

check compose star star

section

parameters {obC obD : Type} {morC : obC → obC → Type} {morD : obD → obD → Type}
 (C : category obC morC)
parameter ( D : category obD morD )

instance C

-- check @id
-- check C
-- check id2

-- inductive functor :=
-- functor_mk : Π (obF : obC → obD) (morF : Π{A B}, morC A B → morD (obF A) (obF B)),
--             (Π {A : obC}, morF (id2 A) = id2 (obF A)) →
-- --            (Π {A B C : obC} {f : morC A B} {g : morC B C}, morF (g ∘ f) = morF g ∘ morF f) →
--             functor

end

-- check @functor_mk
-- check @functor_rec

section
--need extensionality
definition type_cat : category Type (λA B, A → B) :=
cat_mk (λ a b c f g, function.compose f g) (λ a, function.id) (λ a b c d f g h, sorry)
  (λ a b f, sorry) (λ a b f, sorry)

end


end category
