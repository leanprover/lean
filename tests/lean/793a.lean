import data.rat
open rat

check (4.0:rat)
check (2.3:rat)
check (1.00:rat)
check (10.213:rat)

check -(0.3:rat)
check (10.213:rat)
check (2.3:rat)
check (1.0:rat)

eval (Î» v, (1.0:rat)) (2:nat)
