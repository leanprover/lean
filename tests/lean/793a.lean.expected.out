4 : ℚ
23 / 10 : ℚ
1 : ℚ
10213 / 1000 : ℚ
-(3 / 10) : ℚ
10213 / 1000 : ℚ
23 / 10 : ℚ
1 : ℚ
quot.mk (prerat.mk (int.of_nat 1) (int.of_nat 1) (int.of_nat_succ_pos 0))
